src/policy/gc_work.rs

```diff
@@ -0,0 +1,146 @@
+use crate::plan::Plan;
+use crate::plan::TransitiveClosure;
+use crate::policy::space::Space;
+use crate::scheduler::gc_work::*;
+use crate::scheduler::GCWork;
+use crate::scheduler::GCWorker;
+use crate::util::copy::CopySemantics;
+use crate::util::Address;
+use crate::util::ObjectReference;
+use crate::vm::VMBinding;
+use crate::MMTK;
+
+use std::ops::{Deref, DerefMut};
+
+/// Used to identify the trace if a policy has different kinds of traces. For example, defrag vs fast trace for Immix,
+/// mark vs forward trace for mark compact.
+pub(crate) type TraceKind = u8;
+
+/// A plan that uses `PolicyProcessEdges` needs to provide an implementation for this trait.
+/// The plan needs to specify a target space (which needs to implement `SupportPolicyProcessEdges`).
+/// For objects in the target space, `trace_object_with_tracekind()` is called to trace the object.
+/// Otherwise, `fallback_trace()` is used.
+/// For the sake of performance, the implementation of this trait should mark methods as `[inline(always)]`.
+pub trait UsePolicyProcessEdges<VM: VMBinding>: Plan<VM = VM> + Send {
+    type TargetPolicy: SupportPolicyProcessEdges<VM>;
+    /// The copy semantics for objects in the space.
+    const COPY: CopySemantics;
+
+    /// Returns a reference to the target space.
+    fn get_target_space(&self) -> &Self::TargetPolicy;
+
+    /// How to trace objects if the object is not in the default space.
+    fn fallback_trace<T: TransitiveClosure>(
+        &self,
+        trace: &mut T,
+        object: ObjectReference,
+        worker: &mut GCWorker<VM>,
+    ) -> ObjectReference;
+}
+
+/// A policy that allows using `PolicyProcessEdges` needs to provide an implementation for this trait.
+/// For the sake of performance, the implementation of this trait should mark methods as `[inline(always)]`.
+pub trait SupportPolicyProcessEdges<VM: VMBinding>: Space<VM> {
+    /// Trace an object in the policy.
+    fn trace_object_with_tracekind<T: TransitiveClosure, const KIND: TraceKind>(
+        &self,
+        trace: &mut T,
+        object: ObjectReference,
+        copy: CopySemantics,
+        worker: &mut GCWorker<VM>,
+    ) -> ObjectReference;
+
+    /// Create scan work for the policy. By default, we use [`ScanObjects`](crate::scheduler::gc_work::ScanObjects).
+    /// If a policy has its own scan object work packet, they can override this method.
+    #[inline(always)]
+    fn create_scan_work<E: ProcessEdgesWork<VM = VM>>(
+        &'static self,
+        nodes: Vec<ObjectReference>,
+    ) -> Box<dyn GCWork<VM>> {
+        Box::new(crate::scheduler::gc_work::ScanObjects::<E>::new(
+            nodes, false,
+        ))
+    }
```

Comment:
I have made some changes to this:
* add `create_scan_work()` in `ProcessEdgesWork`.
* rename `new_scan_work()` in `ProcessEdgesWork` to `start_scan_work()` to avoid confusion.
* remove the override `flush()` impl in `PolicyProcessEdges`, and instead override `create_scan_work()` and forward the call to `SupportPolicyProcessEdges`.